src/sinks/util/auto_concurrency/service.rs

```diff
@@ -0,0 +1,319 @@
+            } else {
+                assert_pending!(self.service.poll_ready());
+            }
+            Send {
+                request,
+                response,
+                sequence: self.sequence,
+            }
+        }
+
+        fn inner(&self) -> MutexGuard<Inner> {
+            self.inner.lock().unwrap()
+        }
+    }
+
+    impl Send {
+        async fn respond(self) {
+            let data = format!("RESPONSE #{}", self.sequence);
+            self.response.send_response(data.clone());
+            assert_eq!(self.request.await.unwrap(), data);
+        }
+
+        async fn defer(self) {
+            self.response.send_error(TestError::Deferral);
+            assert_downcast_matches!(
+                self.request.await.unwrap_err(),
+                TestError,
+                TestError::Deferral
+            );
+        }
+    }
+
+    #[tokio::test]
+    async fn startup_conditions() {
+        TestService::run(|mut svc| async move {
+            // Concurrency starts at 1
+            assert_eq!(svc.inner().current_limit, 1);
+            svc.send(false).await;
+        })
+        .await;
+    }
+
+    #[tokio::test]
+    async fn increases_limit() {
+        let stats = TestService::run(|mut svc| async move {
+            // Concurrency starts at 1
+            assert_eq!(svc.inner().current_limit, 1);
+            let req = svc.send(false).await;
+            advance(Duration::from_secs(1)).await;
+            req.respond().await;
+
+            // Concurrency stays at 1 until a measurement
+            assert_eq!(svc.inner().current_limit, 1);
+            let req = svc.send(false).await;
+            advance(Duration::from_secs(1)).await;
+            req.respond().await;
+
+            // After a constant speed measurement, concurrency is increased
+            assert_eq!(svc.inner().current_limit, 2);
+        })
+        .await;
+
+        let in_flight = stats.in_flight.stats().unwrap();
+        assert_eq!(in_flight.max, 1);
+        assert_eq!(in_flight.mean, 1.0);
+
+        let observed_rtt = stats.observed_rtt.stats().unwrap();
+        assert_eq!(observed_rtt.mean, 1.0);
+    }
+
+    #[tokio::test]
+    async fn handles_deferral() {
+        TestService::run(|mut svc| async move {
+            assert_eq!(svc.inner().current_limit, 1);
+            let req = svc.send(false).await;
+            advance(Duration::from_secs(1)).await;
+            req.respond().await;
+
+            assert_eq!(svc.inner().current_limit, 1);
+            let req = svc.send(false).await;
+            advance(Duration::from_secs(1)).await;
+            req.respond().await;
+
+            assert_eq!(svc.inner().current_limit, 2);
+
+            let req = svc.send(true).await;
+            advance(Duration::from_secs(1)).await;
+            req.defer().await;
+            assert_eq!(svc.inner().current_limit, 1);
+        })
+        .await;
+    }
+
+    #[allow(clippy::needless_range_loop)]
```

Comment:
I don't think we have a policy, so it would be a useful discussion to have elsewhere. My thinking is to leave as many clippy lints on as possible, to allow all of the code to continue to benefit from them, and turn them off in specific places where we have hand verified they are incorrect. FTR I would like to avoid this one too, but couldn't figure out a way to do it without a range loop (still discussing in Slack though).